src/System.Memory/tests/Performance/Perf.Span.BinarySearch.cs

```diff
@@ -0,0 +1,153 @@
+        [InlineData(1)]
+        [InlineData(10)]
+        [InlineData(100)]
+        [InlineData(1000)]
+        public void SpanBinarySearch_Int_LastIndex(int size)
+        {
+            BenchmarkAndAssert(size, size - 1, size - 1);
+        }
+
+        [Benchmark(InnerIterationCount = InnerCount)]
+        [InlineData(1)]
+        [InlineData(10)]
+        [InlineData(100)]
+        [InlineData(1000)]
+        public void SpanBinarySearch_Int_NotFoundBefore(int size)
+        {
+            BenchmarkAndAssert(size, -1, -1);
+        }
+
+        [Benchmark(InnerIterationCount = InnerCount)]
+        [InlineData(1)]
+        [InlineData(10)]
+        [InlineData(100)]
+        [InlineData(1000)]
+        public void SpanBinarySearch_Int_NotFoundAfter(int size)
+        {
+            BenchmarkAndAssert(size, size, ~size);
+        }
+
+
+        [Benchmark(InnerIterationCount = InnerCount)]
+        [InlineData(1)]
+        [InlineData(10)]
+        [InlineData(100)]
+        [InlineData(1000)]
+        public void SpanBinarySearch_String_FirstIndex(int size)
+        {
+            BenchmarkAndAssert(size, 0.ToString(), 0);
+        }
+
+        [Benchmark(InnerIterationCount = InnerCount)]
+        [InlineData(1)]
+        [InlineData(10)]
+        [InlineData(100)]
+        [InlineData(1000)]
+        public void SpanBinarySearch_String_MiddleIndex(int size)
+        {
+            BenchmarkAndAssert(size, (size / 2).ToString(), size / 2);
+        }
+
+        [Benchmark(InnerIterationCount = InnerCount)]
+        [InlineData(1)]
+        [InlineData(10)]
+        [InlineData(100)]
+        [InlineData(1000)]
+        public void SpanBinarySearch_String_LastIndex(int size)
+        {
+            BenchmarkAndAssert(size, (size - 1).ToString(), size - 1);
+        }
+
+        [Benchmark(InnerIterationCount = InnerCount)]
+        [InlineData(1)]
+        [InlineData(10)]
+        [InlineData(100)]
+        [InlineData(1000)]
+        public void SpanBinarySearch_String_NotFoundBefore(int size)
+        {
+            BenchmarkAndAssert(size, (-1).ToString(), -1);
+        }
+
+        [Benchmark(InnerIterationCount = InnerCount)]
+        [InlineData(1)]
+        [InlineData(10)]
+        [InlineData(100)]
+        [InlineData(1000)]
+        public void SpanBinarySearch_String_NotFoundAfter(int size)
+        {
+            BenchmarkAndAssert(size, (size).ToString(), ~size);
+        }
+
+        private static void BenchmarkAndAssert(int size, int value, int expectedIndex)
+        {
+            Span<int> span = new int[size];
+            for (int i = 0; i < span.Length; i++)
+            {
+                span[i] = i;
+            }
+
+            int index = 0;
+            foreach (BenchmarkIteration iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                {
+                    index |= span.BinarySearch(value);
```

Comment:
You aren't using InnerIterationCount here (or elsewhere). If it isn't needed (i.e. your test results are already > 1 ms), feel free to remove it. Otherwise, I think you need to add a loop within the measured block, using the InnerIterationCount.
See the following for an example - https://github.com/Microsoft/xunit-performance#authoring-benchmarks

As an aside, can you please share the results of the perf tests? I am curious to see them.

Reply:
@ahsonkhan fixed to use `InnerIterationCount, however, `testResults.xml` does not contain any results for these, maybe I am not executing it correctly:
```xml
<?xml version="1.0" encoding="utf-8"?>
<assemblies>
  <assembly name="System.Memory.Performance.Tests.dll" environment="64-bit .NET (unknown version) [collection-per-class, parallel (4 threads)]" test-framework="xUnit.net 2.2.0.3300" run-date="2017-12-11" run-time="21:08:22" total="2" passed="2" failed="0" skipped="0" time="0.221" errors="0">
    <errors />
    <collection total="2" passed="2" failed="0" skipped="0" name="Test collection for System.Buffers.Text.Tests.Base64TestHelper" time="0.014">
      <test name="System.Buffers.Text.Tests.Base64TestHelper.GenerateEncodingMapAndVerify" type="System.Buffers.Text.Tests.Base64TestHelper" method="GenerateEncodingMapAndVerify" time="0.0112869" result="Pass" />
      <test name="System.Buffers.Text.Tests.Base64TestHelper.GenerateDecodingMapAndVerify" type="System.Buffers.Text.Tests.Base64TestHelper" method="GenerateDecodingMapAndVerify" time="0.0028464" result="Pass" />
    </collection>
  </assembly>
</assemblies>
```

Reply:
https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/performance-tests.md does not seem to be correct, at least running:
```
msbuild /t:BuildAndTest /p:Performance=true /p:ConfigurationGroup=Release /p:TargetOS=Windows_NT
```
does not work. What am I doing wrong? 🙄

Reply:
The performance results should be somewhere in this directory: "corefx\bin\tests\System.Memory.Performance.Tests\netcoreapp-Windows_NT-Release-x64", not in testResults.xml

Go to "\corefx\src\System.Memory\tests\Performance" and run the command you listed above:
`msbuild /t:BuildAndTest /p:Performance=true /p:ConfigurationGroup=Release /p:TargetOS=Windows_NT`

Are you getting any errors when you do this?

Reply:
@ahsonkhan no I do not get any errors when running the `msbuild` command, but no tests are run.

Reply:
@ahsonkhan didn't notice this at first but the `msbuild` run says:
```
Skipping target "RunTestsForProject" because all output files are up-to-date with respect to the input files.
```
I thought that weird, and since the `testResults.xml` file was incorrect, only had two tests, that actually do not seem to be proper perf tests, I deleted the file and ran the `msbuild` command again. Then the tests are run and I noted that:
```
[13-12-2017 10:24:05][WRN] Skipping 2 Xunit [Fact]s because they are not [Benchmark]s
```
Therefore I think there are two tests that are wrongly marked as `Benchmark` or something which mess the whole process up.

There is an issue with the string perf tests, so results will have to come later... I think I made a mistake when doing the int to string stuff:
```
[13-12-2017 10:27:55][ERR] System.Memory.Tests.Perf_Span_BinarySearch.SpanBinarySearch_String_NotFoundBefore(size: 100): Assert.Equal() Failure
  Expected: -1
  Actual:   -3
  [13-12-2017 10:27:55][ERR]    at Xunit.Assert.Equal[T](T expected, T actual, IEqualityComparer`1 comparer) in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.assert\Asserts\EqualityAsserts.cs:line 35
     at System.Memory.Tests.Perf_Span_BinarySearch.BenchmarkAndAssert(Int32 size, String value, Int32 expectedIndex) in D:\oss\corefx\src\System.Memory\tests\Performance\Perf.Span.BinarySearch.cs:line 157
     at System.Memory.Tests.Perf_Span_BinarySearch.SpanBinarySearch_String_NotFoundBefore(Int32 size) in D:\oss\corefx\src\System.Memory\tests\Performance\Perf.Span.BinarySearch.cs:line 103
```

Reply:
Followup, weird the actual perf test run does not produce a `testResults.xml` file at the end, maybe an intermediate? Anyway, seems I can run the perf tests after fixing that.

Regarding the possible wrong tests, these are defined as:
```csharp
        [Fact]
        public static void GenerateEncodingMapAndVerify()
        {
            var data = new byte[64]; // Base64
            for (int i = 0; i < s_characters.Length; i++)
            {
                data[i] = (byte)s_characters[i];
            }
            Assert.True(s_encodingMap.AsSpan().SequenceEqual(data));
        }

        [Fact]
        public static void GenerateDecodingMapAndVerify()
        {
            var data = new sbyte[256]; // 0 to byte.MaxValue (255)
            for (int i = 0; i < data.Length; i++)
            {
                data[i] = s_invalidByte;
            }
            for (int i = 0; i < s_characters.Length; i++)
            {
                data[s_characters[i]] = (sbyte)i;
            }
            Assert.True(s_decodingMap.AsSpan().SequenceEqual(data));
        }
```
these are the ones populating the `testResults.xml`. These are static methods on a static class, perhaps there is an issue around that here.